custom/ilsgateway/tasks.py

```diff
@@ -364,3 +368,67 @@ def third_soh_task():
     last_month_last_day, fifth_business_day = get_last_and_nth_business_day(now, 5)
     if now.day == fifth_business_day.day:
         send_for_all_domains(last_month_last_day, SOHReminder)
+
+
+def recalculate_on_group_change(location, last_run):
+    OrganizationSummary.objects.filter(location_id=location.get_id).delete()
+    process_facility_warehouse_data(location, default_start_date(), last_run.end)
+
+    for parent in location.sql_location.get_ancestors(ascending=True):
+        process_non_facility_warehouse_data(parent.couch_location,
+                                            default_start_date(), last_run.end, strict=False)
+
+
+def recalculate_on_parent_change(location, previous_parent_id, last_run):
+    previous_parent = SQLLocation.objects.get(location_id=previous_parent_id)
+    type_location_map = defaultdict(set)
+
+    previous_ancestors = list(previous_parent.get_ancestors(include_self=True, ascending=True))
+    actual_ancestors = list(location.sql_location.get_ancestors(ascending=True))
+
+    locations_to_recalculate = set()
+
+    i = 0
+    while previous_ancestors[i] != actual_ancestors[i] and i < len(previous_ancestors):
```

Comment:
i think there's still a problem here though - what if `len(actual_ancestors)` is shorter than `len(previous_ancestors)`, or if the hierarchies don't match up level-wise - then we wouldn't get a match.

i think it would be better to define a function called `get_common_ancestor` which takes `previous_ancestors` and `actual_ancestors` as arguments and then returns their common ancestor or None if they don't have one.  then you can loop over `previous_ancestors` and keep adding them to `locations_to_recalculate`, and break once you hit the common ancestor, and do the same for `actual_ancestors`. 

and `get_common_ancestor` would be simple - it would just loop over `previous_ancestors` and if an element is `in actual_ancestors` then you return it, otherwise return None if no match is found.